### ISSUE TITLE
[Bug]: Alter table ADD COLUMN NOT NULL with default value does not set null flag correctly.

### ISSUE THREAD

**Issue by fengttt** [matrixorigin/matrixone#16078]
### Is there an existing issue for the same bug?

- [X] I have checked the existing issues.

### Branch Name

main， 1.2-dev

### Commit ID

62bc77f5f

### Other Environment Information

```Markdown
- Hardware parameters:
- OS type:
- Others:
```

### Actual Behavior

```
mysql> alter table t add column yy text not null default 'foo';
Query OK, 0 rows affected (0.01 sec)

mysql> desc t
    -> ;
+-------+---------------+------+------+---------+-------+---------+
| Field | Type          | Null | Key  | Default | Extra | Comment |
+-------+---------------+------+------+---------+-------+---------+
| id    | INT(32)       | NO   | PRI  | NULL    |       |         |
| i     | INT(32)       | YES  |      | NULL    |       |         |
| j     | INT(32)       | YES  |      | NULL    |       |         |
| t     | TEXT(0)       | YES  |      | NULL    |       |         |
| e     | ENUM('Y','N') | YES  |      | 'Y'     |       |         |
| x     | INT(32)       | NO   |      | NULL    |       |         |
| y     | TEXT(0)       | NO   |      | NULL    |       |         |
| yy    | TEXT(0)       | YES  |      | 'foo'   |       |         |
+-------+---------------+------+------+---------+-------+---------+
8 rows in set (0.01 sec)
```

### Expected Behavior

Note that column yy should be NOT NULL, that is, Null column should be NO.

Aslo I checked that I can insert a value into t with yy as null.

alter table modify column does not work either.

```mysql> insert into t values (3, 3, 3, 'foo', null, 2, 'bar', null);
Query OK, 1 row affected (0.01 sec)

mysql> select * from t;
+------+------+------+------+------+------+------+------+
| id   | i    | j    | t    | e    | x    | y    | yy   |
+------+------+------+------+------+------+------+------+
|    3 |    3 |    3 | foo  | NULL |    2 | bar  | NULL |
|    1 |    1 |    1 | foo  | NULL |    0 |      | foo  |
+------+------+------+------+------+------+------+------+
2 rows in set (0.01 sec)

mysql>
```

### Steps to Reproduce

```Markdown
create table t (id int not null primary key, i int, j int, t text);

alter table t add column e enum('Y', 'N') not null default 'Y';

desc t;
```

**Comment by qingxinhome** [comment]
The issue has been resolved, please verify and test， @fengttt @aressu1985 @heni02 @daviszhen

**Comment by aronchanisme** [comment]
Fixed in both `1.2-dev` and `main`. Hi @Ariznawlll could you please kindly help add some bvt cases? Thanks a lot.

**Test case:**
```sql
create database if not exists test;
use test;
drop table if exists t;
create table t (id int not null primary key, i int, j int, t text);
desc t;
alter table t add column e enum('Y', 'N') not null default 'Y';
desc t;
```

**Test result:**
1. `1.2-dev`(`18f806e26`)
```bash
github@test0-10-222-4-0:/data$ mo_ctl start
2024-05-27 14:24:51.401 UTC+0800    [INFO]    No mo-service is running
2024-05-27 14:24:51.449 UTC+0800    [DEBUG]    Check total memory on current machine, command: free -m | awk 'NR==2{print }', result(Mi): 15996
2024-05-27 14:24:51.498 UTC+0800    [INFO]    Get conf succeeded: MO_DEPLOY_MODE="git"
2024-05-27 14:24:51.540 UTC+0800    [INFO]    GO memory limit(Mi): 9597
2024-05-27 14:24:51.575 UTC+0800    [DEBUG]    Start command will add GOMEMLIMIT=9597MiB
2024-05-27 14:24:51.610 UTC+0800    [INFO]    Starting mo-service: cd /data/mo/1.2-dev/matrixone/ && GOMEMLIMIT=9597MiB /data/mo/1.2-dev/matrixone/mo-service -daemon -debug-http :9876 -launch /data/mo/1.2-dev/matrixone/etc/launch/launch.toml >/data/mo/1.2-dev/matrixone/logs/stdout-20240527_142451.log 2>/data/mo/1.2-dev/matrixone/logs/stderr-20240527_142451.log
2024-05-27 14:24:51.782 UTC+0800    [INFO]    Wait for 2 seconds
2024-05-27 14:24:53.881 UTC+0800    [INFO]    At least one mo-service is running. Process info: 
github   1593384       1 23 14:24 ?        00:00:00 /data/mo/1.2-dev/matrixone/mo-service -daemon -debug-http :9876 -launch /data/mo/1.2-dev/matrixone/etc/launch/launch.toml
2024-05-27 14:24:53.917 UTC+0800    [INFO]    List of pid(s): 
1593384
2024-05-27 14:24:53.952 UTC+0800    [INFO]    Start succeeded
github@test0-10-222-4-0:/data$ mo_ctl get_cid
2024-05-27 14:25:23.539 UTC+0800    [INFO]    Try get mo commit id
commit 18f806e26a516a2990d1b348728d7f838a5ba43b
Author: gouhongshen <gouhongshen@hotmail.com>
Date:   Sat May 25 12:21:13 2024 +0800

    Revert "sync to 1.2: remove the old fast ranges and update metrics (#… (#16401)
    
    …16359)"
    
    This reverts commit 918a3569df85d9dfe02d5ddfa04a3e15352d04f9.
    
    ## What type of PR is this?
    
    - [ ] API-change
    - [ ] BUG
    - [ ] Improvement
    - [ ] Documentation
    - [ ] Feature
    - [ ] Test and CI
    - [ ] Code Refactoring
    
    ## Which issue(s) this PR fixes:
    
    issue #https://github.com/matrixorigin/matrixone/issues/16379
    
    ## What this PR does / why we need it:
    revert to avoid oom
2024-05-27 14:25:23.595 UTC+0800    [INFO]    Get commit id succeeded
github@test0-10-222-4-0:/data$ mo_ctl get_branch
2024-05-27 14:25:25.799 UTC+0800    [INFO]    Try get mo branch
2024-05-27 14:25:25.853 UTC+0800    [INFO]    Get branch succeeded, current branch: 1.2-dev
github@test0-10-222-4-0:/data$ mo_ctl connect
2024-05-27 14:25:28.981 UTC+0800    [INFO]    Checking connectivity
2024-05-27 14:25:29.057 UTC+0800    [INFO]    Ok, connecting for user ... 
Welcome to the MySQL monitor.  Commands end with ; or \g.
Your MySQL connection id is 8
Server version: 8.0.30-MatrixOne-v1880626 MatrixOne

Copyright (c) 2000, 2024, Oracle and/or its affiliates.

Oracle is a registered trademark of Oracle Corporation and/or its
affiliates. Other names may be trademarks of their respective
owners.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

mysql> create database if not exists test;
Query OK, 1 row affected (0.02 sec)

mysql> use test;
Database changed
mysql> create table t (id int not null primary key, i int, j int, t text);
Query OK, 0 rows affected (0.02 sec)

mysql> desc t;
+-------+---------+------+------+---------+-------+---------+
| Field | Type    | Null | Key  | Default | Extra | Comment |
+-------+---------+------+------+---------+-------+---------+
| id    | INT(32) | NO   | PRI  | NULL    |       |         |
| i     | INT(32) | YES  |      | NULL    |       |         |
| j     | INT(32) | YES  |      | NULL    |       |         |
| t     | TEXT(0) | YES  |      | NULL    |       |         |
+-------+---------+------+------+---------+-------+---------+
4 rows in set (0.01 sec)

mysql> alter table t add column e enum('Y', 'N') not null default 'Y';
Query OK, 0 rows affected (0.04 sec)

mysql> desc t;
+-------+---------------+------+------+---------+-------+---------+
| Field | Type          | Null | Key  | Default | Extra | Comment |
+-------+---------------+------+------+---------+-------+---------+
| id    | INT(32)       | NO   | PRI  | NULL    |       |         |
| i     | INT(32)       | YES  |      | null    |       |         |
| j     | INT(32)       | YES  |      | null    |       |         |
| t     | TEXT(0)       | YES  |      | null    |       |         |
| e     | ENUM('Y','N') | NO   |      | 'Y'     |       |         |
+-------+---------------+------+------+---------+-------+---------+
5 rows in set (0.01 sec)

mysql> select git_version();
+---------------+
| git_version() |
+---------------+
| 18f806e26     |
+---------------+
1 row in set (0.00 sec)

mysql> exit
Bye
```

2. `main`(`28682fdf9`)
```bash
github@test0-10-222-4-0:/data$ mo_ctl start 
2024-05-27 14:32:02.574 UTC+0800    [INFO]    No mo-service is running
2024-05-27 14:32:02.623 UTC+0800    [DEBUG]    Check total memory on current machine, command: free -m | awk 'NR==2{print }', result(Mi): 15996
2024-05-27 14:32:02.674 UTC+0800    [INFO]    Get conf succeeded: MO_DEPLOY_MODE="git"
2024-05-27 14:32:02.716 UTC+0800    [INFO]    GO memory limit(Mi): 9597
2024-05-27 14:32:02.752 UTC+0800    [DEBUG]    Start command will add GOMEMLIMIT=9597MiB
2024-05-27 14:32:02.788 UTC+0800    [INFO]    Starting mo-service: cd /data/mo/main/matrixone/ && GOMEMLIMIT=9597MiB /data/mo/main/matrixone/mo-service -daemon -debug-http :9876 -launch /data/mo/main/matrixone/etc/launch/launch.toml >/data/mo/main/matrixone/logs/stdout-20240527_143202.log 2>/data/mo/main/matrixone/logs/stderr-20240527_143202.log
2024-05-27 14:32:02.981 UTC+0800    [INFO]    Wait for 2 seconds
2024-05-27 14:32:05.076 UTC+0800    [INFO]    At least one mo-service is running. Process info: 
github   1598779       1 24 14:32 ?        00:00:00 /data/mo/main/matrixone/mo-service -daemon -debug-http :9876 -launch /data/mo/main/matrixone/etc/launch/launch.toml
2024-05-27 14:32:05.112 UTC+0800    [INFO]    List of pid(s): 
1598779
2024-05-27 14:32:05.148 UTC+0800    [INFO]    Start succeeded
\github@test0-10-222-4-0:/data$ mo_ctl get_branch
2024-05-27 14:50:41.310 UTC+0800    [INFO]    Try get mo branch
2024-05-27 14:50:41.367 UTC+0800    [INFO]    Get branch succeeded, current branch: main
github@test0-10-222-4-0:/data$ mo_ctl get_cid
2024-05-27 14:50:44.142 UTC+0800    [INFO]    Try get mo commit id
commit 28682fdf988e1906bef3b869b13702c6790ea6bd
Author: Kai Cao <ck89119@users.noreply.github.com>
Date:   Mon May 27 14:05:22 2024 +0800

    fix get fk table failed when doing restoration (#16408)
2024-05-27 14:50:44.203 UTC+0800    [INFO]    Get commit id succeeded
github@test0-10-222-4-0:/data$ mo_ctl connect
2024-05-27 14:50:46.978 UTC+0800    [INFO]    Checking connectivity
2024-05-27 14:50:47.056 UTC+0800    [INFO]    Ok, connecting for user ... 
Welcome to the MySQL monitor.  Commands end with ; or \g.
Your MySQL connection id is 30
Server version: 8.0.30-MatrixOne-v286829 MatrixOne

Copyright (c) 2000, 2024, Oracle and/or its affiliates.

Oracle is a registered trademark of Oracle Corporation and/or its
affiliates. Other names may be trademarks of their respective
owners.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

mysql> select git_version();
+---------------+
| git_version() |
+---------------+
| 28682fdf9     |
+---------------+
1 row in set (0.00 sec)

mysql> create database if not exists test;
Query OK, 1 row affected (0.02 sec)

mysql> use test;
Database changed
mysql> create table t (id int not null primary key, i int, j int, t text);
Query OK, 0 rows affected (0.02 sec)

mysql> desc t;
+-------+---------+------+------+---------+-------+---------+
| Field | Type    | Null | Key  | Default | Extra | Comment |
+-------+---------+------+------+---------+-------+---------+
| id    | INT(32) | NO   | PRI  | NULL    |       |         |
| i     | INT(32) | YES  |      | NULL    |       |         |
| j     | INT(32) | YES  |      | NULL    |       |         |
| t     | TEXT(0) | YES  |      | NULL    |       |         |
+-------+---------+------+------+---------+-------+---------+
4 rows in set (0.02 sec)

mysql> alter table t add column e enum('Y', 'N') not null default 'Y';
Query OK, 0 rows affected (0.04 sec)

mysql> desc t;
+-------+---------------+------+------+---------+-------+---------+
| Field | Type          | Null | Key  | Default | Extra | Comment |
+-------+---------------+------+------+---------+-------+---------+
| id    | INT(32)       | NO   | PRI  | NULL    |       |         |
| i     | INT(32)       | YES  |      | null    |       |         |
| j     | INT(32)       | YES  |      | null    |       |         |
| t     | TEXT(0)       | YES  |      | null    |       |         |
| e     | ENUM('Y','N') | NO   |      | 'Y'     |       |         |
+-------+---------------+------+------+---------+-------+---------+
5 rows in set (0.01 sec)
```

**Comment by Ariznawlll** [comment]
**case:**
```
create table t (id int not null primary key, i int, j int, t text);
alter table t add column e enum('Y', 'N') not null default 'Y';
desc t;
alter table t modify column i int not null;
desc t;
```
**main: 28682fdf988e1906bef3b869b13702c6790ea6bd**
<img width="602" alt="image" src="https://github.com/matrixorigin/matrixone/assets/108530700/1e1451b5-fa80-4ca1-abee-0914d3708fc1">

**1.2-dev: 28682fdf988e1906bef3b869b13702c6790ea6bd**
<img width="574" alt="image" src="https://github.com/matrixorigin/matrixone/assets/108530700/d6f3cbe2-b789-4468-87f2-93a4dd4ae49e">

verified.
